### ISSUE TITLE
Can't save attribute via REST API - LogicException: Property "IsUsedInGrid" does not have corresponding setter in class "Magento\Catalog\Api\Data\ProductAttributeInterface"

### ISSUE THREAD

**Issue by careys7** [magento/magento2#8827]
### Preconditions

1. Magento 2.1.5, PHP 7.0

### Steps to reproduce

1. Send Web API request as follows:

```
PUT /rest/default/V1/products/attributes/short_description/ HTTP/1.1
Host: example.com
Authorization: OAuth oauth_consumer_key="xxxxx",oauth_token="xxxxx",oauth_signature_method="HMAC-SHA1",oauth_timestamp="xxxxx",oauth_nonce="xxxxx",oauth_signature="xxxxx"
Content-Type: application/json
Cache-Control: no-cache

{
  "attribute": {
    "is_wysiwyg_enabled": true,
    "is_html_allowed_on_front": true,
    "used_for_sort_by": false,
    "is_filterable": false,
    "is_filterable_in_search": false,
    "is_used_in_grid": true,
    "is_visible_in_grid": false,
    "is_filterable_in_grid": false,
    "position": 0,
    "apply_to": [],
    "is_searchable": "1",
    "is_visible_in_advanced_search": "1",
    "is_comparable": "1",
    "is_used_for_promo_rules": "0",
    "is_visible_on_front": "0",
    "used_in_product_listing": "1",
    "is_visible": true,
    "scope": "store",
    "attribute_id": 76,
    "attribute_code": "short_description",
    "frontend_input": "textarea",
    "entity_type_id": "4",
    "is_required": false,
    "options": [],
    "is_user_defined": false,
    "default_frontend_label": "Short Description",
    "backend_type": "text",
    "is_unique": "0",
    "validation_rules": []
  }
}
```

### Expected result

1. Attribute updated successfully

### Actual result

1. Error returned via API:
```
{
  "message": "Internal Error. Details are available in Magento log file. Report ID: webapi-58c0d36765f2b"
}
```

Contents of `webapi-58c0d36765f2b`:

```
[2017-03-09 04:00:39] report.CRITICAL: LogicException: Property "IsUsedInGrid" does not have corresponding setter in class "Magento\Catalog\Api\Data\ProductAttributeInterface". in /var/www/html/vendor/magento/framework/Reflection/NameFinder.php:100
Stack trace:
#0 /var/www/html/vendor/magento/framework/Reflection/NameFinder.php(74): Magento\Framework\Reflection\NameFinder->findAccessorMethodName(Object(Zend\Code\Reflection\ClassReflection), 'IsUsedInGrid', 'setIsUsedInGrid', 'setIsIsUsedInGr...')
#1 /var/www/html/vendor/magento/framework/Webapi/ServiceInputProcessor.php(163): Magento\Framework\Reflection\NameFinder->getSetterMethodName(Object(Zend\Code\Reflection\ClassReflection), 'IsUsedInGrid')
#2 /var/www/html/vendor/magento/framework/Webapi/ServiceInputProcessor.php(322): Magento\Framework\Webapi\ServiceInputProcessor->_createFromArray('Magento\\Catalog...', Array)
#3 /var/www/html/vendor/magento/framework/Webapi/ServiceInputProcessor.php(119): Magento\Framework\Webapi\ServiceInputProcessor->convertValue(Array, 'Magento\\Catalog...')
#4 /var/www/html/vendor/magento/module-webapi/Controller/Rest/InputParamsResolver.php(101): Magento\Framework\Webapi\ServiceInputProcessor->process('Magento\\Catalog...', 'save', Array)
#5 /var/www/html/vendor/magento/module-webapi/Controller/Rest.php(299): Magento\Webapi\Controller\Rest\InputParamsResolver->resolve()
#6 /var/www/html/vendor/magento/module-webapi/Controller/Rest.php(216): Magento\Webapi\Controller\Rest->processApiRequest()
#7 /var/www/html/vendor/magento/framework/Interception/Interceptor.php(146): Magento\Webapi\Controller\Rest->dispatch(Object(Magento\Framework\App\Request\Http))
#8 /var/www/html/var/generation/Magento/Webapi/Controller/Rest/Interceptor.php(39): Magento\Webapi\Controller\Rest\Interceptor->___callPlugins('dispatch', Array, Array)
#9 /var/www/html/vendor/magento/framework/App/Http.php(135): Magento\Webapi\Controller\Rest\Interceptor->dispatch(Object(Magento\Framework\App\Request\Http))
#10 /var/www/html/vendor/magento/framework/App/Bootstrap.php(258): Magento\Framework\App\Http->launch()
#11 /var/www/html/pub/index.php(37): Magento\Framework\App\Bootstrap->run(Object(Magento\Framework\App\Http))
#12 {main}

Next Exception: Report ID: webapi-58c0d36765f2b; Message: Property "IsUsedInGrid" does not have corresponding setter in class "Magento\Catalog\Api\Data\ProductAttributeInterface". in /var/www/html/vendor/magento/framework/Webapi/ErrorProcessor.php:195
Stack trace:
#0 /var/www/html/vendor/magento/framework/Webapi/ErrorProcessor.php(139): Magento\Framework\Webapi\ErrorProcessor->_critical(Object(LogicException))
#1 /var/www/html/vendor/magento/module-webapi/Controller/Rest.php(219): Magento\Framework\Webapi\ErrorProcessor->maskException(Object(LogicException))
#2 /var/www/html/vendor/magento/framework/Interception/Interceptor.php(146): Magento\Webapi\Controller\Rest->dispatch(Object(Magento\Framework\App\Request\Http))
#3 /var/www/html/var/generation/Magento/Webapi/Controller/Rest/Interceptor.php(39): Magento\Webapi\Controller\Rest\Interceptor->___callPlugins('dispatch', Array, Array)
#4 /var/www/html/vendor/magento/framework/App/Http.php(135): Magento\Webapi\Controller\Rest\Interceptor->dispatch(Object(Magento\Framework\App\Request\Http))
#5 /var/www/html/vendor/magento/framework/App/Bootstrap.php(258): Magento\Framework\App\Http->launch()
#6 /var/www/html/pub/index.php(37): Magento\Framework\App\Bootstrap->run(Object(Magento\Framework\App\Http))
#7 {main} [] []
```

**Comment by lepetit** [comment]
I have the same issue with Magento 2.1.6

I noticed a substantial difference between the two similar call:

## **/rest/default/V1/products/attributes/{attributeId}**
In this case the returned object is fully populated, with all the fields, including **is_used_in_grid**

```
{
  "is_wysiwyg_enabled": false,
  "is_html_allowed_on_front": true,
  "used_for_sort_by": true,
  "is_filterable": false,
  "is_filterable_in_search": false,
  "is_used_in_grid": true,
  "is_visible_in_grid": true,
  "is_filterable_in_grid": true,
  "position": 0,
  "apply_to": [],
  "is_searchable": "1",
  "is_visible_in_advanced_search": "0",
  "is_comparable": "0",
  "is_used_for_promo_rules": "0",
  "is_visible_on_front": "0",
  "used_in_product_listing": "1",
  "is_visible": true,
  "scope": "global",
  "attribute_id": 144,
  "attribute_code": "brand",
  "frontend_input": "select",
  "entity_type_id": "4",
  "is_required": false,
  "options": [
    {
      "label": " ",
      "value": ""
    }
  ],
  "is_user_defined": true,
  "default_frontend_label": "Brand",
  "frontend_labels": null,
  "backend_type": "int",
  "source_model": "Magento\\Eav\\Model\\Entity\\Attribute\\Source\\Table",
  "default_value": "",
  "is_unique": "0",
  "validation_rules": []
}

```
## **/V1/products/attribute-sets/{attributeSetId}/attributes**
In this case the full list of defined attributes is returned, with much less data for each:

```
{
   "attribute_id":144,
   "attribute_code":"brand",
   "frontend_input":"select",
   "entity_type_id":"4",
   "is_required":false,
   "options":[
      {
         "label":" ",
         "value":""
      }
   ],
   "is_user_defined":true,
   "default_frontend_label":"Brand",
   "frontend_labels":null,
   "backend_type":"int",
   "source_model":"Magento\\Eav\\Model\\Entity\\Attribute\\Source\\Table",
   "default_value":"",
   "is_unique":"0",
   "validation_rules":[   ]
}
```

As a workaround I use this code:
```

            List<CatalogDataProductAttributeInterface> attributes = manager.CatalogProductAttributeManagementV1GetAttributesGet("4");
            foreach (CatalogDataProductAttributeInterface attribute in attributes)
            {
                if (attribute.AttributeCode == "brand")
                {
                    EavDataAttributeOptionInterface option = new EavDataAttributeOptionInterface("Text", "Value");
                    attribute.Options.Add(option);

                    CatalogProductAttributeRepositoryV1Api attributeRepository = new CatalogProductAttributeRepositoryV1Api(config);
                    Body20 body = new Body20(attribute);
                    CatalogDataProductAttributeInterface result= attributeRepository.CatalogProductAttributeRepositoryV1SavePut(attribute.AttributeCode, body);                    
                }
            }

```
It is c# but you can easily adapt to any other language.

I hope this helps,
Nicola

**Comment by RomanKrut** [comment]
@careys7 Hi! Thanks for reporting this issue. The internal ticket was created MAGETWO-70475.

**Comment by magento-engcom-team** [comment]
@careys7, thank you for your report.
The issue is already fixed in develop branch, 2.2.0
